File: consorcio/lib/infrastructure/firestore/firebase_support_ticket_repository.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../../domain/entities/support_ticket.dart';
import '../../domain/errors/domain_exception.dart';
import '../../domain/repositories/support_ticket_repository.dart';

class FirebaseSupportTicketRepository implements SupportTicketRepository {
  FirebaseSupportTicketRepository({FirebaseFirestore? firestore})
      : _firestore = firestore ?? FirebaseFirestore.instance;

  final FirebaseFirestore _firestore;

  CollectionReference<Map<String, dynamic>> get _tickets =>
      _firestore.collection('supportTickets');

  SupportTicket _map(String id, Map<String, dynamic> data) {
    return SupportTicket(
      id: id,
      kind: (data['kind'] as String?) ?? 'SUGERENCIA',
      message: (data['message'] as String?) ?? '',
      status: (data['status'] as String?) ?? 'ABIERTO',
      createdById: (data['createdById'] as String?) ?? '',
      createdByName: (data['createdByName'] as String?) ?? '',
      createdAt: (data['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      response: (data['response'] as String?) ?? '',
      resolvedAt: (data['resolvedAt'] as Timestamp?)?.toDate(),
      resolvedById: (data['resolvedById'] as String?) ?? '',
      resolvedByName: (data['resolvedByName'] as String?) ?? '',
    );
  }

  @override
  Future<SupportTicket> create({
    required String kind,
    required String message,
    required String createdById,
    required String createdByName,
  }) async {
    final payload = {
      'kind': kind,
      'message': message,
      'status': 'ABIERTO',
      'createdById': createdById,
      'createdByName': createdByName,
      'createdAt': FieldValue.serverTimestamp(),
      'response': '',
      'resolvedById': '',
      'resolvedByName': '',
    };
    final doc = await _tickets.add(payload);
    final snapshot = await doc.get();
    return _map(doc.id, snapshot.data() ?? payload);
  }

  @override
  Future<List<SupportTicket>> listMine(String userId) async {
    final snapshot = await _tickets
        .where('createdById', isEqualTo: userId)
        .orderBy('createdAt', descending: true)
        .limit(80)
        .get();
    return snapshot.docs.map((doc) => _map(doc.id, doc.data())).toList();
  }

  @override
  Future<List<SupportTicket>> listAll() async {
    final snapshot =
        await _tickets.orderBy('createdAt', descending: true).limit(200).get();
    return snapshot.docs.map((doc) => _map(doc.id, doc.data())).toList();
  }

  @override
  Future<SupportTicket> resolve({
    required String ticketId,
    required String response,
    required String resolvedById,
    required String resolvedByName,
  }) async {
    final ref = _tickets.doc(ticketId);
    final existing = await ref.get();
    if (!existing.exists) {
      throw DomainException('El aviso no existe');
    }
    await ref.update({
      'status': 'RESUELTO',
      'response': response,
      'resolvedAt': FieldValue.serverTimestamp(),
      'resolvedById': resolvedById,
      'resolvedByName': resolvedByName,
    });
    final updated = await ref.get();
    return _map(ticketId, updated.data()!);
  }
}

File: src/infrastructure/firestore/FirebaseSupportTicketRepository.ts
import {
  collection,
  doc,
  getDoc,
  getDocs,
  orderBy,
  query,
  Timestamp,
  updateDoc,
} from 'firebase/firestore'
import type { SupportTicket } from '@/domain/entities/SupportTicket'
import type {
  ResolveSupportTicketInput,
  SupportTicketRepository,
} from '@/domain/repositories/SupportTicketRepository'
import { NotFoundError } from '@/domain/errors/DomainError'
import { firestoreDb } from '@/infrastructure/firebase/firebaseApp'

interface SupportTicketDoc {
  kind: 'SUGERENCIA' | 'PROBLEMA'
  message: string
  status: 'ABIERTO' | 'RESUELTO'
  createdById: string
  createdByName: string
  createdAt: Timestamp
  response?: string
  resolvedAt?: Timestamp | null
  resolvedById?: string
  resolvedByName?: string
}

function mapTicket(id: string, data: SupportTicketDoc): SupportTicket {
  return {
    id,
    kind: data.kind === 'PROBLEMA' ? 'PROBLEMA' : 'SUGERENCIA',
    message: data.message ?? '',
    status: data.status === 'RESUELTO' ? 'RESUELTO' : 'ABIERTO',
    createdById: data.createdById ?? '',
    createdByName: data.createdByName ?? '',
    createdAt: data.createdAt?.toDate?.() ?? new Date(),
    response: data.response ?? '',
    resolvedAt: data.resolvedAt?.toDate?.() ?? null,
    resolvedById: data.resolvedById ?? '',
    resolvedByName: data.resolvedByName ?? '',
  }
}

export class FirebaseSupportTicketRepository implements SupportTicketRepository {
  private readonly collectionRef = collection(firestoreDb, 'supportTickets')

  async listAll(): Promise<SupportTicket[]> {
    const snapshot = await getDocs(
      query(this.collectionRef, orderBy('createdAt', 'desc')),
    )
    return snapshot.docs.map((item) =>
      mapTicket(item.id, item.data() as SupportTicketDoc),
    )
  }

  async resolve(input: ResolveSupportTicketInput): Promise<SupportTicket> {
    const ref = doc(this.collectionRef, input.ticketId)
    const existing = await getDoc(ref)
    if (!existing.exists()) {
      throw new NotFoundError('Aviso no encontrado')
    }
    await updateDoc(ref, {
      status: 'RESUELTO',
      response: input.response,
      resolvedAt: Timestamp.now(),
      resolvedById: input.resolvedById,
      resolvedByName: input.resolvedByName,
    })
    const updated = await getDoc(ref)
    return mapTicket(input.ticketId, updated.data() as SupportTicketDoc)
  }
}
